### PR TITLE
Transformer optimizations

### PR DESCRIPTION
**Pull request by tbennun**
This PR includes two major modifications to transformers:
* Batching heads in multi-head attention into single operations instead of on a per-head basis
* Stacking the weights and biases for queries/keys/values in self-attention

The PR also includes fixes for the weighted sum operator (w.r.t. in-place layers) and path fixes for the transformer test

### CODE REVIEW

File: python/lbann/models/transformer.py [hunk 1]
```diff
@@ -393,6 +396,7 @@ def __init__(
             )
             for i in range(num_decoder_layers)
         ]
+        self.separate_heads = self.decoder[0].attention1.separate_heads
```
Comment (ndryden) [inline python/lbann/models/transformer.py:399]: Is it worth sanity-checking that all layers have the same `separate_heads` setting?

File: python/lbann/modules/transformer.py [hunk 2]
```diff
@@ -91,26 +108,118 @@ def forward(self, queries, keys, values, mask=None):
...
+            # Unstack
+            qkv_slice = lbann.Slice(qkv_fc,
+                                    axis=1,
+                                    slice_points=[
+                                        0, self.embed_dim, 2 * self.embed_dim,
+                                        3 * self.embed_dim
+                                    ])
+            queries_fc = lbann.Identity(qkv_slice)
+            keys_fc = lbann.Identity(qkv_slice)
+            values_fc = lbann.Identity(qkv_slice)
```
Comment (ndryden) [inline python/lbann/modules/transformer.py:133]: Not a criticism, but we should make a note that this notation for slicing is quite unintuitive. Maybe we should revisit it at some point.

**Review by ndryden — APPROVED** [review]
This all looks correct to me. Nice work!

File: python/lbann/models/transformer.py (hunk 1 above)
Comment (tbennun) [reply python/lbann/models/transformer.py:399]: Done

**Review by bvanessen — APPROVED** [review]

**Pull request merged**